File: 125_algorithms/_exercises/templates/_algorithms_challenges/algorithm-master/leetcode/36_valid_sudoku.py
c_ Solution:
    ___ isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        __ n.. board o. n.. board[0] o. l..(board) != l..(board[0]):
            r.. F..

        n = l..(board)
        EMPTY = '.'
        CANDS = '123456789'

        ___ x __ r..(n):
            used = set()

            ___ y __ r..(n):
                __ board[x][y] __ EMPTY:
                    continue
                __ board[x][y] n.. __ CANDS:
                    r.. F..
                __ board[x][y] __ used:
                    r.. F..
                used.add(board[x][y])

        ___ y __ r..(n):
            used = set()

            ___ x __ r..(n):
                __ board[x][y] __ EMPTY:
                    continue
                __ board[x][y] __ used:
                    r.. F..
                used.add(board[x][y])

        ___ i __ r..(3):
            ___ j __ r..(3):
                used = set()

                ___ x __ r..(i * 3, i * 3 + 3):
                    ___ y __ r..(j * 3, j * 3 + 3):
                        __ board[x][y] __ EMPTY:
                            continue
                        __ board[x][y] __ used:
                            r.. F..
                        used.add(board[x][y])

        r.. T..

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/HashTable/36_ValidSudoku.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-17 10:15:37
 */

c.. Solution
{
public:
    /* Hash Table
    Use three 9*9 array to keep the used numbers in row, col, panel.
    According to:
    https://discuss.leetcode.com/topic/8241/my-short-solution-by-c-o-n2
    */
    bool isValidSudoku(vector<vector<char> > &board)
    {
        int row[9][9] = {0}, col[9][9] = {0}, panel[9][9] = {0};

        ___(int i = 0; i < board.size(); ++i)
            ___(int j = 0; j < board[i].size(); ++j)
                __(board[i][j] != '.')
                {
                    int num = board[i][j] - '0' - 1, k = i / 3 * 3 + j / 3;
                    __(row[i][num] || col[j][num] || panel[k][num])
                        r_ false;
                    row[i][num] = col[j][num] = panel[k][num] = 1;
                }

        r_ true;
    }
};

/*
["..4...63.",".........","5......9.","...56....","4.3.....1","...7.....","...5.....",".........","........."]
[".87654321","2........","3........","4........","5........","6........","7........","8........","9........"]
*/
